File: code/java/src/main/java/flyweight/ConcreteBook.java
package flyweight;

class ConcreteBook implements Book {
    private final String title;
    private final String author;
    private boolean isAvailable = true;
    private int id;

    public ConcreteBook(String title, String author, int id){
        this.title = title;
        this.author = author;
        this.id = id;

    }

    @Override
    public void borrowBook() {
        if (isAvailable) {
            isAvailable = false;
            System.out.println("Book borrowed: " + title + " by " + author + " book id: " + id);
        } else {
            System.out.println("Book is currently not available: " + title);
        }
    }

    @Override
    public void returnBook() {
        isAvailable = true;
        System.out.println("Book returned: " + title + " by " + author + " book id: " + id);
    }

    @Override
    public boolean isAvailable() {
        return isAvailable;
    }

    public String getTitle() {
        return title;
    }

    @Override
    public int getId() {
        return id;
    }
}

File: code/typescript/src/flyweight/ConcreteBook.ts
import { Book } from "./Book";

export class ConcreteBook implements Book {
    private available = true;

    constructor(private title: string, private author: string, private id: number) { }

    borrowBook(): void {
        if (this.available) {
            this.available = false;
            console.log(`Book borrowed: ${this.title} by ${this.author} book id: ${this.id}`);
        } else {
            console.log(`Book is currently not available: ${this.title}`);
        }
    }

    returnBook(): void {
        this.available = true;
        console.log(`Book returned: ${this.title} by ${this.author} book id: ${this.id}`);
    }

    isAvailable(): boolean {
        return this.available;
    }

    getId(): number {
        return this.id;
    }
}
